secrets/certregistry/certregistry.go

```diff
@@ -0,0 +1,218 @@
+package certregistry
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	defaultHost = "ingress.local"
+	currentTime = time.Now
+	// ErrNoMatchingCertificateFound is used if there is no matching certificate found
+	ErrNoMatchingCertificateFound = errors.New("no matching certificate found")
+)
+
+const (
+	// used as wildcard char in Cert Hostname/AltName matches
+	glob = "*"
+	// minimal time period for the NotAfter attribute of a Cert to be in the future
+	minimalCertValidityPeriod = 7 * 24 * time.Hour
+)
+
+type tlsCertificate struct {
+	hosts []string
+	cert  *tls.Certificate
+}
+
+// CertRegistry object holds TLS certificates to be used to terminate TLS connections
+// We ensure ensure syncronized access to them and hold a default certificate.
+type CertRegistry struct {
+	lookup         map[string]*tlsCertificate
+	mx             *sync.Mutex
+	defaultTLSCert *tls.Certificate
+}
+
+// NewCertRegistry initializes the certificate registry with an empty map
+// and a generated default certificate.
+func NewCertRegistry() *CertRegistry {
+	l := make(map[string]*tlsCertificate)
+
+	return &CertRegistry{
+		lookup:         l,
+		mx:             &sync.Mutex{},
+		defaultTLSCert: getFakeHostTLSCert(defaultHost),
+	}
+}
+
+func (r *CertRegistry) getCertByKey(key string) (*tlsCertificate, bool) {
+	r.mx.Lock()
+	defer r.mx.Unlock()
+
+	cert, ok := r.lookup[key]
+	if !ok || cert == nil {
+		log.Debugf("certificate not found in registry - %s", key)
+		return nil, false
+	}
+
+	return cert, true
+}
+
+func (r *CertRegistry) addCertToRegistry(key string, cert *tlsCertificate) {
+	r.mx.Lock()
+	defer r.mx.Unlock()
+
+	r.lookup[key] = cert
+}
+
+// SyncCert takes a TLS certificate and list of hosts and saves them to the registry with the
+// provided key. If the object already exists it will be updated or added otherwise. Returns
+// true if key was changed.
+func (r *CertRegistry) SyncCert(key string, hosts []string, crt *tls.Certificate) bool {
```

Comment:
Why do we need to know if cert was changed? I don't see the result being used.
If we would check this then it should be done properly with mutex locked, otherwise there is a race condition between getting and adding the cert

Reply:
Why do we need key here at all? It looks like some magic value that turns out to be secret name+namespace. Maybe we should simply keep `map[string]*tls.Certificate` mappinf and use host a a key?

Reply:
Updated to use the suggested register and lookup method with a simple `map[string]*tls.Certificate`